### PR TITLE
update pocketpy from v1.x to v2.0.1

### PR DESCRIPTION
**Pull request by PrimedErwin**
Update pocketpy to v2.0.1

- Changed submodule pocketpy to [pockepy v2](https://github.com/PrimedErwin/pocketpy/tree/main).
- Updated apis in src/api/python.c using pkpy v2.
- Updated cmake to avoid some compile problems.
- APIs are tested (might still be bugs here).

### CODE REVIEW

File: src/api/python.c [hunk 1]
```diff
@@ -1,48 +1,63 @@
 #include "core/core.h"
 
-#include "pocketpy_c.h"
 #include <stdio.h>
 #include <string.h>
+#include "pocketpy.h"
+#include "pocketpy/objects/base.h"
```
Comment (blueloveTH) [inline src/api/python.c:6]: DO NOT include internal headers.

File: src/api/python.c [hunk 2]
```diff
@@ -1,48 +1,63 @@
...
-struct CachedNames{
-    pkpy_CName _tic_core;
-    pkpy_CName len;
-    pkpy_CName __getitem__;
-    pkpy_CName TIC;
-    pkpy_CName BOOT;
-    pkpy_CName SCN;
-    pkpy_CName BDR;
-    pkpy_CName MENU;
+struct names
```
Comment (blueloveTH) [inline src/api/python.c:18]: Please keep the old name `CachedNames`.

File: src/api/python.c [hunk 3]
```diff
@@ -1,48 +1,63 @@
...
+    py_Name BOOT;
+    py_Name SCN;
+    py_Name BDR;
+    py_Name MENU;
 } N;
 
-static void pkpy_setglobal_2(pkpy_vm* vm, const char* name){
-    pkpy_setglobal(vm, pkpy_name(name));
+//set value of "name" to stack.top
+static bool pkpy_setglobal_2(const char* name)
```
Comment (blueloveTH) [inline src/api/python.c:31]: Delete useless function.

File: src/api/python.c [hunk 4]
```diff
@@ -1,48 +1,63 @@
...
+{
+    py_setglobal(py_name(name), py_peek(-1));
 }
 
-static bool get_core(pkpy_vm* vm, tic_core** core)
+static bool get_core(tic_core** core)
```
Comment (blueloveTH) [inline src/api/python.c:36]: Change this function to `tic_core* get_core()` and it must return a valid pointer.

File: src/api/python.c [hunk 5]
```diff
@@ -1,48 +1,63 @@
...
+    return true;
 }
 
-static bool setup_core(pkpy_vm* vm, tic_core* core)
+//set _tic_core to the given core
+static bool setup_core(tic_core* core)
 {
-    if (!pkpy_push_voidp(vm, core)) return false;
-    return pkpy_setglobal(vm, N._tic_core);
+    py_newint(py_retval(), (py_i64)core);
```
Comment (blueloveTH) [inline src/api/python.c:47]: Use `py_r0()` instead of `py_retval()`.

File: src/api/python.c [hunk 6]
```diff
@@ -54,1579 +69,1273 @@ static int prepare_colorindex(pkpy_vm* vm, int index, u8 * buffer)
...
-        pkpy_vectorcall(vm, 1);
-
-        int list_len;
-        pkpy_to_int(vm, -1, &list_len);
-        pkpy_pop_top(vm);
-
-        list_len = (list_len < TIC_PALETTE_SIZE)?(list_len):(TIC_PALETTE_SIZE);
-
-        for(int i = 0; i < list_len; i++)
+        int list_len = py_list_len(index);
```
Comment (blueloveTH) [inline src/api/python.c:72]: Call `py_checktype(..., tp_list)` before using any function of list.

**Review by blueloveTH — CHANGES_REQUESTED** [review]

File: src/api/python.c [hunk 7]
```diff
@@ -576,7 +576,7 @@ static bool py_mouse(int argc, py_Ref argv)
     tic_point pos = core->api.mouse(tic);
     const tic80_mouse* mouse = &core->memory.ram->input.mouse;
 
-    py_Ref res = py_retval();
+    py_Ref res = (py_Ref)malloc(sizeof(py_Ref)*7);
```
Comment (blueloveTH) [inline src/api/python.c:579]: This is incorrect.

File: src/api/python.c [hunk 8]
```diff
@@ -576,7 +576,8 @@ static bool py_mouse(int argc, py_Ref argv)
     tic_point pos = core->api.mouse(tic);
     const tic80_mouse* mouse = &core->memory.ram->input.mouse;
 
-    py_Ref res = (py_Ref)malloc(sizeof(py_Ref)*7);
+    //py_Ref res = (py_Ref)malloc(sizeof(py_Ref));
+    py_Ref res = py_r7();
     py_newtuple(res, 7);
     py_Ref tmp = py_retval();
     py_newint(tmp, pos.x);
```
Comment (blueloveTH) [inline src/api/python.c:583]: Use one line instead of two.
```c
py_newint(py_tuple_getitem(0), pos.x)
```

**Review by blueloveTH — CHANGES_REQUESTED** [review]
Should take a look at `CMakeLists.txt`.

**Review by blueloveTH — APPROVED** [review]

**Review by nesbox — APPROVED** [review]
Thank you

**Pull request merged**